### PR TITLE
custom inference code

### PR DESCRIPTION
**Pull request by ArneRobben**
## What does this change?

This PR adds a custom inference script that uses the same multi-label thresholding as applied to the logits in training.

## How to test

```python -m stage_model stage```
Beware you need an environment file with the following:
```
WANDB_ENTITY=wellcome-ml-team
WANDB_API_KEY=****
MODEL_REGISTRY=wellcome-ml-team/model-registry/HRCSTagger
SAGEMAKER_BUCKET=****
SAGEMAKER_USER_CONFIG_OVERRIDE=****
```
The last two locations can be changed to any internal S3 location
```python src/deploy.py``` 
Beware that for me this didn't work from my EC2 (sagemaker credentials restriction) so had to be run locally

## How can we measure success?

We need to run this past DataOps to flag the output can now contain more than one label

## Have we considered potential risks?
- [X]  By testing the above, endpoints were created on Sagemaker, these are now all deleted
- [ ]  There might be some integration issues with Salesforce as a result, we need to check this doesn't cause any issues

### CODE REVIEW

File: src/stage_model.py [hunk 1]
```diff
@@ -41,19 +41,37 @@ def add_tokenizer(artifact_dir):
...
+def add_custom_inference_script(artifact_dir):
+    """Add custom inference script to the artifact directory.
+    Args:
+        artifact_dir (str): Path to the artifact directory.
+    """
+    code_dir = os.path.join(artifact_dir, "code")
+    os.makedirs(code_dir, exist_ok=True)
+
+    # Read the contents of your existing inference.py
+    with open("src/inference.py", "r") as src_file:
```
Comment (Amvla) [inline src/stage_model.py]: Feel free to leave it as is, but just to note this will assume you're in your main directory. Alternatively, if you want something more generic you could use something like
```
dirname = os.path.dirname(__file__)
filename = os.path.join(dirname, 'inference.py')
```

File: src/stage_model.py [hunk 2]
```diff
@@ -41,19 +41,37 @@ def add_tokenizer(artifact_dir):
...
 
+def add_custom_inference_script(artifact_dir):
+    """Add custom inference script to the artifact directory.
+    Args:
+        artifact_dir (str): Path to the artifact directory.
+    """
+    code_dir = os.path.join(artifact_dir, "code")
+    os.makedirs(code_dir, exist_ok=True)
+
+    # Read the contents of your existing inference.py
```
Comment (Amvla) [inline src/stage_model.py]: This function is straightforward enough to understand that the inline comments can be safely removed.

File: src/inference.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        predictions_name = []
+        predictions_probabilities = []
+        for i in range(predictions.shape[0]):
+            pred_indices = np.where(predictions[i] == 1)[0]
+            pred_names = [model.config.id2label[idx] for idx in pred_indices]
+            pred_probs = [logits[i][idx] for idx in pred_indices]
+            predictions_name.append(pred_names)
+            predictions_probabilities.append(pred_probs)
+
+        results = []
```
Comment (Amvla) [inline src/inference.py]: Instead of storing prediction names and probabilities first in lists and then creating your final results dictionary, you could insert them directly into your dictionary inside your first loop:
```
results = []
for i in range(predictions.shape[0]):
    pred_indices = np.where(predictions[i] == 1)[0]
    pred_names = [model.config.id2label[idx] for idx in pred_indices]
    pred_probs = [logits[i][idx] for idx in pred_indices]
    results.append({
        "label": pred_names,
        "score": pred_probs
    })
```

File: src/inference.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        
+
+
+
+        # now only keep predictions and logits for the predictions that are 1
+        predictions_name = []
+        predictions_probabilities = []
+        for i in range(predictions.shape[0]):
+            pred_indices = np.where(predictions[i] == 1)[0]
+            pred_names = [model.config.id2label[idx] for idx in pred_indices]
```
Comment (Amvla) [inline src/inference.py]: This reminds me - have you added a custom `id2label` into the config as part of this change? I'm just asking because previously it was just `LABEL_0` etc., which is fine but Jeff is using a mapping json that I shared to get the correct labels. So if this is not needed anymore he'd need to be notified.

**Review by Amvla — COMMENTED** [review]
Overall looks good, it's nice that adding a custom `inference.py` into the huggingface container does the trick! 🎉
I've left some comments, nothing too involved. Also just to add a slightly picky comment, I was wondering if you can use a different directory to save the model artifact to? `sagemaker_test` might be a bit misleading here given that this is the actual artifact that the production DataOps endpoint is going to be pointing to. Ideally we'd want to set this to a directory that is more safe from accidental deletion.

File: src/stage_model.py (hunk 1 above)
Comment (ArneRobben) [reply src/stage_model.py]: nice

File: src/inference.py (hunk 4 above)
Comment (ArneRobben) [reply src/inference.py]: the labels are now added in `train.py`

File: src/stage_model.py [hunk 5]
```diff
@@ -41,19 +41,36 @@ def add_tokenizer(artifact_dir):
...
+    Args:
+        artifact_dir (str): Path to the artifact directory.
+    """
+    code_dir = os.path.join(artifact_dir, "code")
+    os.makedirs(code_dir, exist_ok=True)
+
+    dirname = os.path.dirname(__file__)
+    filename = os.path.join(dirname, "inference.py")
+
+    # use shutil to copy the file
```
Comment (Amvla) [inline src/stage_model.py]: I think this comment can be removed

File: src/inference.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    text_input = data["inputs"]
+    text_input = " ".join(text_input.split())
+
+    inputs = tokenizer(text_input, return_tensors="pt", truncation=True, padding=True)
+    with torch.no_grad():
+        outputs = model(**inputs)
+        logits = outputs.logits
+        logits = torch.sigmoid(torch.tensor(logits)).cpu().detach().numpy()
+
+        thresholds = [1] * logits.shape[1]
```
[the remark this answers is not in the document]
Comment (ArneRobben) [reply src/inference.py]: This is really cool, I like it
